datahugger/services.py

```diff
@@ -445,3 +451,113 @@ class B2shareDataset(DatasetDownloader):
     ATTR_SIZE_JSONPATH = "size"
     ATTR_HASH_JSONPATH = "checksum"
     ATTR_HASH_TYPE_VALUE = "md5"
+
+
+class YodaDataset(DatasetDownloader):
+    """Downloader for Yoda repositories."""
+
+    def _get_contents_url(self):
+        """Resolve the root folder of the contents of a Yoda data package."""
+        url_to_use = (
+            self.resource if isinstance(self.resource, str) else self.resource.resolve()
+        )
+        res = requests.get(url_to_use)
+        res.raise_for_status()
+        soup = BeautifulSoup(res.content, "html.parser")
+        contents_link = soup.find("a", string=re.compile(r"View contents"))
+        return contents_link.get("href") + "/original" if contents_link else None
+
+    @property
+    def files(self):
+        if not hasattr(self, "_files"):
+            self._requests_cache_file = tempfile.NamedTemporaryFile(delete=False)
+            requests_cache.install_cache(self._requests_cache_file.name)
+            self._files = self._harvest_files()
+            self._cleanup_requests_cache()
+        return self._files
+
+    def _cleanup_requests_cache(self):
+        if hasattr(self, "_requests_cache_file"):
+            if os.path.isfile(self._requests_cache_file.name):
+                self._requests_cache_file.close()
+                os.unlink(self._requests_cache_file.name)
+
+    def _get_collection_name_from_folder_url(self, base_url, folder_url):
+        result = folder_url.replace(base_url, "", 1)
+        return result[1:] if result.startswith("/") else result
+
+    def _get_full_url(self, base_url, relative_collection, object_name):
+        result = base_url
+        if relative_collection != "":
+            result = urljoin(
+                result if result.endswith("/") else result + "/", relative_collection
+            )
+        result = urljoin(result if result.endswith("/") else result + "/", object_name)
+        return result
+
+    def _get_relative_path(self, collection_name, dataobject_name):
+        separator = (
+            "/" if collection_name != "" and not collection_name.endswith("/") else ""
+        )
+        return f"{collection_name}{separator}{dataobject_name}"
+
+    def _harvest_files(self):
+        contents_url = self._get_contents_url()
+        if contents_url is None:
+            raise ValueError(
+                "Data package contents link not found. "
+                + "This can happen if the Yoda data package is not open access.\n"
+            )
+
+        folders_to_process = [contents_url]
+        files_to_download = []
+
+        while True:
```

Comment:
I wonder whether the while loop is needed here.

Reply:
The purpose of the loop is to iterate through any subcollections (subdirectories) of the data packages if needed. I've adjusted the loop condition and added a comment to make this clearer.